src/main/java/org/opentripplanner/graph_builder/module/OsmBoardingLocationsModule.java

```diff
@@ -0,0 +1,146 @@
+
+  @Override
+  public void buildGraph(
+    Graph graph,
+    HashMap<Class<?>, Object> extra,
+    DataImportIssueStore issueStore
+  ) {
+    var streetIndex = graph.getStreetIndex();
+    LOG.info("Improving boarding locations by checking OSM entities...");
+    int successes = 0;
+    for (TransitStopVertex ts : graph.getVerticesOfType(TransitStopVertex.class)) {
+      // if the street is already linked there is no need to linked it again,
+      // could happened if using the prune isolated island
+      boolean alreadyLinked = false;
+      for (Edge e : ts.getOutgoing()) {
+        if (e instanceof StreetTransitStopLink) {
+          alreadyLinked = true;
+          break;
+        }
+      }
+      if (alreadyLinked) continue;
+      // only connect transit stops that are not part of a pathway network
+      if (!ts.hasPathways()) {
+        if (!connectVertexToStop(ts, streetIndex, graph.getLinker())) {
+          LOG.debug(
+            "Could not connect " + ts.getStop().getCode() + " at " + ts.getCoordinate().toString()
+          );
+        } else {
+          successes++;
+        }
+      }
+    }
+    LOG.info("Found {} OSM references which match a stop's id or code", successes);
+  }
+
+  @Override
+  public void checkInputs() {
+    //no inputs
+  }
+
+  private boolean connectVertexToStop(
+    TransitStopVertex ts,
+    StreetVertexIndex index,
+    VertexLinker linker
+  ) {
+    var stopCode = ts.getStop().getCode();
+    var stopId = ts.getStop().getId().getId();
+    Envelope envelope = new Envelope(ts.getCoordinate());
+
+    double xscale = Math.cos(ts.getCoordinate().y * Math.PI / 180);
+    envelope.expandBy(searchRadiusDegrees / xscale, searchRadiusDegrees);
+    var vertices = index
+      .getVerticesForEnvelope(envelope)
+      .stream()
+      .filter(OsmBoardingLocationVertex.class::isInstance)
+      .map(OsmBoardingLocationVertex.class::cast)
+      .toList();
+    // Iterate over all nearby vertices representing transit stops in OSM, linking to them if they have a stop code or di
+    // in their ref= tag that matches the GTFS stop code of this StopVertex.
+    for (var osmVertex : vertices) {
+      if (
+        (stopCode != null && osmVertex.references.contains(stopCode)) ||
+        osmVertex.references.contains(stopId)
+      ) {
+        if (!osmVertex.isConnectedToStreetNetwork()) {
+          linker.linkVertexPermanently(
+            osmVertex,
+            new TraverseModeSet(TraverseMode.WALK),
+            LinkingDirection.BOTH_WAYS,
+            (osmBoardingLocationVertex, splitVertex) -> {
+              // the OSM boarding location vertex is not connected to the street network, so we
+              // need to link it to the platform
+              return List.of(
+                linkBoardingLocationToStreetNetwork(osmVertex, splitVertex),
+                linkBoardingLocationToStreetNetwork(splitVertex, osmVertex)
+              );
```

Comment:
Our dataset now displays a few links with zero length.  I guess this is not a fault in this code but in the linker. My interpretation is that the computed platform average point often lies exactly on top of visibility edges, and this leads to collapsed links. They seem not to cause any harm in routing, though.,

Reply:
This is something that I want to tackle next. 

Right now I'm using the regular linker to link the centroid to the street network. However, OTP should consider the boarding location part of the platform and make it part of the visibility calculation.

This will also improve the quality of the linking in difficult cases like Five Points station in Atlanta, where the platforms are on top of each other.

- https://www.openstreetmap.org/relation/7915137
- https://www.openstreetmap.org/relation/7915138